metricbeat/module/mssql/db/db.go

```diff
@@ -0,0 +1,176 @@
+package db
+
+import (
+	"database/sql"
+	"fmt"
+	_ "github.com/denisenkom/go-mssqldb"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/module/mssql"
+	"github.com/pkg/errors"
+	"sync"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "db", New, mb.DefaultMetricSet())
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*mssql.MetricSet
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The mssql io metricset is experimental.")
+
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	//TODO Db must be gracefully closed
```

Comment:
Not sure if needed here, you probably want to connect and disconnect on `Fetch`
But if you want to start something in `New` you can implement also the [`Closer` interface](https://godoc.org/github.com/elastic/beats/metricbeat/mb#Closer) in the MetricSet to do the cleanup when it is stopped.

Reply:
I was actually thinking how was the lifetime of a metricset and a module

Reply:
I have seen that you have added a `Close` method, but as I said you probably want to connect and disconnect on `Fetch`. The main problem of leaving the connection open during all the life of the metricset is that you need to handle reconnections if the connection is closed by the server or any other external reason. This can add unnecessary complexity.